### PR TITLE
feat: notifications

### PR DESCRIPTION
**Pull request by fgerschwiler**

### CODE REVIEW

File: internal/api/auth/context.go [hunk 1]
```diff
@@ -59,6 +59,10 @@ func VerifyTokenAndWriteCtxData(ctx context.Context, token, orgID string, t Toke
 	return context.WithValue(ctx, dataKey, CtxData{UserID: userID, OrgID: orgID, ProjectID: projectID, AgentID: agentID}), nil
 }
 
+func GetCtxDataKey() key {
```
Comment (fgerschwiler) [inline internal/api/auth/context.go:62]: will be removed as soon as setup is mergef

File: internal/user/repository/view/model/notify_user.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+		Gender:            model.Gender(user.Gender),
+		LastEmail:         user.LastEmail,
+		VerifiedEmail:     user.VerifiedEmail,
+		LastPhone:         user.LastPhone,
+		VerifiedPhone:     user.VerifiedPhone,
+		Sequence:          user.Sequence,
+	}
+}
+
+func (p *NotifyUser) AppendEvent(event *models.Event) (err error) {
```
Comment (adlerhurst) [inline internal/user/repository/view/model/notify_user.go:78]: var name (p *NotifyUser)

File: internal/notification/repository/eventsourcing/spooler/lock.go [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+type lock struct {
+	LockerID    string    `gorm:"column:locker_id;primary_key"`
+	LockedUntil time.Time `gorm:"column:locked_until"`
+	ViewName    string    `gorm:"column:object_type;primary_key"`
+}
+
+func (l *locker) Renew(lockerID, viewModel string, waitTime time.Duration) error {
+	return crdb.ExecuteTx(context.Background(), l.dbClient, nil, func(tx *sql.Tx) error {
+		query := fmt.Sprintf("INSERT INTO %s (%s, %s, %s) VALUES ($1, $2, now()+$3) ON CONFLICT (%s) DO UPDATE SET %s = now()+$4, %s = $5 WHERE (locks.%s < now() OR locks.%s = $6) AND locks.%s = $7",
+			lockTable, objectTypeKey, lockerIDKey, lockedUntilKey, objectTypeKey, lockedUntilKey, lockerIDKey, lockedUntilKey, lockerIDKey, objectTypeKey)
```
Comment (adlerhurst) [inline internal/notification/repository/eventsourcing/spooler/lock.go:33]: maybe we can prepare the query outside the function

File: internal/api/auth/context.go (hunk 1 above)
Comment (adlerhurst) [reply internal/api/auth/context.go]: i think setup is merged

Comment (fgerschwiler) [reply internal/api/auth/context.go]: yes i already did this

File: internal/user/model/phone.go [hunk 4]
```diff
@@ -1,13 +1,18 @@
...
-	"github.com/caos/logging"
 	"github.com/caos/zitadel/internal/crypto"
+	caos_errs "github.com/caos/zitadel/internal/errors"
 	es_models "github.com/caos/zitadel/internal/eventstore/models"
+	"github.com/ttacon/libphonenumber"
 	"time"
 )
 
+const (
+	//TODO: How do we get region?
```
[the remark this answers is not in the document]
Comment (adlerhurst) [reply internal/user/model/phone.go:12]: will be done later

File: internal/notification/providers/email/provider.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+	if err != nil {
+		return err
+	}
+
+	defer logging.LogWithFields("EMAI-a1c87ec8").Debug("email sent")
+	return email.smtpClient.Quit()
+}
+
+func (smtpConfig SMTP) connectToSMTP(tlsRequired bool) (client *smtp.Client, err error) {
+	host, _, _ := net.SplitHostPort(smtpConfig.Host)
```
Comment (livio-a) [inline internal/notification/providers/email/provider.go:72]: please check the error

File: internal/view/failed_events.go [hunk 6]
```diff
@@ -76,7 +76,7 @@ func LatestFailedEvent(db *gorm.DB, table, viewName string, sequence uint64) (*F
 	query := PrepareGetByQuery(table, queries...)
 	err := query(db, failedEvent)
 
-	if err == nil {
+	if err == nil && failedEvent != nil {
```
Comment (livio-a) [inline internal/view/failed_events.go:79]: can't be nil as it's initialised by `new`

File: internal/user/model/phone.go [hunk 7]
```diff
@@ -23,27 +28,17 @@ type PhoneCode struct {
...
-	if err := json.Unmarshal(event.Data, p); err != nil {
-		logging.Log("EVEN-dlo9s").WithError(err).Error("could not unmarshal event data")
-		return err
+func (p *Phone) formatPhone() error {
+	phoneNr, err := libphonenumber.Parse(p.PhoneNumber, defaultRegion)
+	if err != nil {
+		return caos_errs.ThrowPreconditionFailed(nil, "EVENT-so0wa", "Phonenumber is invalid")
 	}
+	p.PhoneNumber = libphonenumber.Format(phoneNr, libphonenumber.E164)
+	libphonenumber.GetSupportedGlobalNetworkCallingCodes()
```
Comment (livio-a) [inline internal/user/model/phone.go:41]: isn't used

File: internal/notification/providers/email/message.go [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+)
+
+var (
+	isHTMLRgx = regexp.MustCompile(`.*<html.*>.*`)
+	lineBreak = "\n"
```
Comment (livio-a) [inline internal/notification/providers/email/message.go:11]: RFC (https://tools.ietf.org/html/rfc2822#section-2.1) states to use CR LF (\r\n)

File: internal/notification/providers/chat/provider.go [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+	"net/url"
+	"unicode/utf8"
+)
+
+type Chat struct {
+	URL        *url.URL
+	SplitCount int
+}
+
+func InitChatProvider(config *ChatConfig) (*Chat, error) {
```
Comment (livio-a) [inline internal/notification/providers/chat/provider.go:18]: i'd remove the pointers in the init funcs as configs are just structs in systemDefaults anyway

File: internal/notification/repository/eventsourcing/spooler/lock.go (hunk 3 above)
Comment (fgerschwiler) [reply internal/notification/repository/eventsourcing/spooler/lock.go]: i moved the whole renew event to es

File: internal/notification/providers/email/provider.go [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+		return caos_errs.ThrowInternal(nil, "EMAIL-s8JLs", "message is not EmailMessage")
+	}
+	// To && From
+	if err := email.smtpClient.Mail(emailMsg.SenderEmail); err != nil {
+		return caos_errs.ThrowInternalf(err, "EMAIL-s3is3", "could not set sender: %v", emailMsg.SenderEmail)
+	}
+
+	for _, recp := range append(append(emailMsg.Recipients, emailMsg.CC...), emailMsg.BCC...) {
+		if err := email.smtpClient.Rcpt(recp); err != nil {
+			return caos_errs.ThrowInternalf(err, "EMAIL-s3is3", "could not set recipient: %v", recp)
```
Comment (adlerhurst) [inline internal/notification/providers/email/provider.go:47]: same error id

Comment (adlerhurst) [reply internal/notification/providers/email/provider.go:47]:
```suggestion
			return caos_errs.ThrowInternalf(err, "EMAIL-s4is4", "could not set recipient: %v", recp)
```

File: internal/notification/repository/eventsourcing/handler/notification.go [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+}
+
+func (n *Notification) handleEmailVerificationCode(event *models.Event) (err error) {
+	alreadyHandled, err := n.checkIfCodeAlreadyHandled(event.AggregateID, event.Sequence, es_model.UserEmailCodeAdded, es_model.UserEmailCodeSent)
+	if err != nil {
+		return err
+	}
+	if alreadyHandled {
+		return nil
+	}
```
Comment (adlerhurst) [inline internal/notification/repository/eventsourcing/handler/notification.go:110]:
```suggestion
	if err != nil || alreadyHandled {
		return nil
	}
```

File: internal/notification/repository/eventsourcing/handler/notification.go [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+}
+
+func (n *Notification) handlePhoneVerificationCode(event *models.Event) (err error) {
+	alreadyHandled, err := n.checkIfCodeAlreadyHandled(event.AggregateID, event.Sequence, es_model.UserPhoneCodeAdded, es_model.UserPhoneCodeSent)
+	if err != nil {
+		return err
+	}
+	if alreadyHandled {
+		return nil
+	}
```
Comment (adlerhurst) [inline internal/notification/repository/eventsourcing/handler/notification.go:131]:
```suggestion
	if err != nil || alreadyHandled {
		return nil
	}
```

Comment (adlerhurst) [reply internal/notification/repository/eventsourcing/handler/notification.go:128]: not done

**Pull request merged**